### PR TITLE
Move rendering of swatches to the server side

### PR DESCRIPTION
**Pull request by krzksz**
### Description (*)

I would like to present you a result of multiple days of work in which I achieved the goal of moving rendering of swatches from JavaScript to PHP.

#### Rationale

Rendering critical page elements in JavaScript is rarely a good idea when the majority of everything else is done on the server side. Not only you get a flash of missing content which results of layout shifting when JavaScript loads but also get a performance hit of modifying the DOM in the very first moments of page initialization. Because of the fact, that the whole configuration for swatch renderer was embedded into the server response as JSON anyway, there were no reasons against porting the whole logic into PHP.

#### Changes summary

- Original PHP API is preserved.
- HTML markup and styling is pretty much identical.
- All of the relevant rendering config was moved to XML.
- Rendering is split into multiple blocks which should make any customizations easier.
- Attribute label is added always, it's just visually hidden when needed so screen readers can still use it.
- JavaScript API is preserved.
- Swatch Renderer Tooltip component was extracted to a separate file so it can be loaded conditionally and inline JavaScript initialization can be removed from aftersearch navigation renderer.
- Hiding of swatches after more button was moved to pure CSS.
- Each SwatchRenderer JS widget now takes 95% less time to initialize.

### Manual testing scenarios (*)

1. Functionality of swatches is preserved in all places they are displayed.

### Questions or comments
Feedback appreciated.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by krzksz** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @krzksz. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @krzksz, here is your [new Magento instance](https://pr-28005.instances.magento-community.engineering).
Admin access: [https://pr-28005.instances.magento-community.engineering/admin_6718](https://pr-28005.instances.magento-community.engineering/admin_6718)
Login: `a7cd2cd9` Password: `e197e47a53b6`
Instance will be terminated in up to 3 hours.

**Comment by DrewML** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @DrewML. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @DrewML, here is your [new Magento instance](https://pr-28005.instances.magento-community.engineering).
Admin access: [https://pr-28005.instances.magento-community.engineering/admin_9f74](https://pr-28005.instances.magento-community.engineering/admin_9f74)
Login: `74795d4d` Password: `2a0c97a9ddaa`
Instance will be terminated in up to 3 hours.

**Comment by krzksz** [comment]
I can see some tests failing but looks like nothing major, I'll fix them ASAP 👍

File: app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+ */
+
+define([
+    'jquery',
+    'jquery-ui-modules/widget'
+], function ($) {
+    'use strict';
+
+    /**
+     * Render tooltips by attributes (only to up).
```
Comment (DrewML) [inline app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js:13]: Can you clarify what `(only to up)` means in this context?

File: app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+define([
+    'jquery',
+    'jquery-ui-modules/widget'
```
Comment (DrewML) [inline app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js:8]: It's so nice to start seeing these granular imports!

File: app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js (hunk 1 above)
Comment (DrewML) [reply app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js:13]: Oh nevermind, just saw that was copypasta from `swatch-renderer.js`. Is this whole file just copy + paste or are there further customizations?

Comment (krzksz) [reply app/code/Magento/Swatches/view/base/web/js/swatch-renderer-tooltip.js:13]: No, this is a simple copy paste.

**Comment by krzksz** [comment]
@magento run all tests

**Review by ptylek — APPROVED** [review]
Great work

**Comment by magento-engcom-team** [comment]
Hi @ptylek, thank you for the review.
[ENGCOM-7653](https://jira.corp.magento.com/browse/ENGCOM-7653) has been created to process this Pull Request
:eight_spoked_asterisk: @ptylek, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by ptylek — CHANGES_REQUESTED** [review]
Can you check failing tests?

**Comment by krzksz** [comment]
@magento run Static Tests,Unit Tests

**Comment by krzksz** [comment]
@magento run Functional Tests CE

**Comment by krzksz** [comment]
@magento run all tests

**Comment by krzksz** [comment]
@magento run Static Tests,Unit Tests,Functional Tests CE

**Comment by krzksz** [comment]
@magento run Static Tests

**Comment by krzksz** [comment]
@magento run all tests

**Comment by krzksz** [comment]
Ready for review I think! 😃

**Review by lbajsarowicz — DISMISSED** [review]
Everything looks fine from my side :-)

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7653](https://jira.corp.magento.com/browse/ENGCOM-7653) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by ptylek — DISMISSED** [review]
Approved. Also mage.swatchRendererTooltip getting some love, I believe it wasn't returned before.

**Comment by magento-engcom-team** [comment]
Hi @ptylek, thank you for the review.
[ENGCOM-7653](https://jira.corp.magento.com/browse/ENGCOM-7653) has been created to process this Pull Request
:eight_spoked_asterisk: @ptylek, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @krzksz With fix configurable product with visual swatch has dropdown with swatch options on category page
**Manual testing scenario:**

- Create configurable product with visual swatch attribute
image
image

- Go to category page

:heavy_check_mark: **Expected result:**
No configurable options is shown
#28005MainVisual

:x: **Actual result:**
Configurable options are shown and displayed as dropdown
#28005PRVisualColor

Could you take a look?

**Comment by engcom-Charlie** [comment]
Hi @krzksz, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Pull request closed without merging**